### PR TITLE
Allow multiple methods for running unit tests

### PR DESCRIPTION
**Pull request by billsacks**
Besides "scripts_regression_tests.py A_RunUnitTests", there are two other methods for running unit tests that I like to use:

1. From scripts/lib, run "python3 -m unittest discover -s CIME/tests -t .": this is similar to what's run by A_RunUnitTests, but doesn't run the doc tests. This runs significantly faster than running "scripts_regression_tests.py A_RunUnitTests".
2. From scripts/lib, run an individual test file, like: "./CIME/tests/test_user_mod_support.py". This runs a single module's tests very fast. I like this method for doing TDD, because I can run unit tests super frequently, in the blink of an eye.

A recent change broke a few unit tests under method (1); this PR fixes that issue. In addition, method (2) wasn't working consistently for the different unit test files; this PR fixes that usage for all of the unit tests under CIME/tests and its subdirectories.

Test suite: scripts_regression_tests.py on cheyenne, and hand-testing of the above alternative options for running unit tests
Test baseline: N/A
Test namelist changes: none
Test status: bit for bit

Fixes none

User interface changes?: N

Update gh-pages html (Y/N)?: N

### CODE REVIEW

File: scripts/lib/CIME/case/case.py
Comment (billsacks) [inline scripts/lib/CIME/case/case.py:87]: A bit of explanation for this change and the related changes in this file – doing away with the recently-added in_doctest() and in_unit_test() functions: I have replaced these functions with explicit handling in the unit tests and doctests that needed it, either by ensuring that caseroot is a non-existent directory or by putting an empty `env_case.xml` file in the caseroot directory. Reasons I prefer this method: (1) (main reason) it works for `python -m unittest discover`; (2) I found the other method hard to understand; (3) it makes me nervous to have unit test-specific code sprinkled in the production code, especially when that test-specific code is hard to understand: it's hard to be confident that things will work right in production.

File: scripts/lib/CIME/case/case.py
Comment (billsacks) [inline scripts/lib/CIME/case/case.py:594]: In addition to the general comment above: For these doctests, it looks like they were creating a case in the current directory (wherever that happens to be), which feels like it could cause trouble in some circumstances, so it felt safer to fix these tests.

File: scripts/lib/CIME/tests/test_utils.py
Comment (billsacks) [inline scripts/lib/CIME/tests/test_utils.py:10]: Relative import doesn't work if you try to run this as a standalone script

File: scripts/lib/CIME/tests/test_provenance.py
Comment (billsacks) [inline scripts/lib/CIME/tests/test_provenance.py:9]: Relative import doesn't work if you try to run this as a standalone script

**Comment by billsacks** [comment]
As noted here https://github.com/ESMCI/cime/pull/4009#pullrequestreview-688928542 there will be conflicts with #4009 in at least a couple of files; we should keep the version in this PR for the conflicted shebang lines.

**Review by jedwards4b — APPROVED** [review]

**Comment by jedwards4b** [comment]
@billsacks Thanks!

**Comment by jedwards4b** [comment]
@jasonb5 Is this still waiting for your review?

**Comment by billsacks** [comment]
I just merged latest master into this and resolved some minor conflicts.

**Comment by billsacks** [comment]
I just merged up to the latest master, resolving conflicts. @jasonb5 I assume that your LGTM comment means you approve even though you didn't formally approve this PR? If I haven't heard from you by, say, mid next week, I'd like to go ahead and merge this.

**Review by jasonb5 — APPROVED** [review]
Forgot to formally approve this, it's good to go though.

**Pull request merged**
